File: frontend/src/domain/provenance.ts
/**
 * Shared provenance model (ADR-010). Embedded on every engineering value or
 * object that can plausibly originate from more than one place, so the
 * measured/imported/user-entered/assumed/calculated distinction is uniform
 * across the whole domain model rather than reinvented per feature.
 */

export type ProvenanceOriginType =
  | "imported"
  | "user-entered"
  | "assumed"
  | "transformed"
  | "interpolated"
  | "calculated"
  | "library-default";

export type VerificationState = "unverified" | "verified" | "rejected";

export interface Provenance {
  readonly originType: ProvenanceOriginType;
  readonly sourceFile?: string;
  readonly sourceFileHash?: string;
  readonly sourceRef?: string;
  /** ISO 8601 */
  readonly importedAt?: string;
  /** ISO 8601 */
  readonly modifiedAt?: string;
  readonly calculationMethod?: string;
  readonly calculationParameters?: Readonly<Record<string, unknown>>;
  readonly softwareVersion?: string;
  readonly verificationState: VerificationState;
  readonly notes?: string;
}

File: backend/app/domain/provenance.py
"""
Mirrors frontend/src/domain/provenance.ts (ADR-010) -- the backend has
never needed to produce provenance-carrying domain objects before (only
point-cloud processing results), so this is the first Python port of it,
for the new PLS-POLE (.pol) import (app/processing/pol_import.py).
"""

from __future__ import annotations

from typing import Literal

from app.schemas.camel_model import CamelModel

ProvenanceOriginType = Literal[
    "imported",
    "user-entered",
    "assumed",
    "transformed",
    "interpolated",
    "calculated",
    "library-default",
]

VerificationState = Literal["unverified", "verified", "rejected"]


class Provenance(CamelModel):
    origin_type: ProvenanceOriginType
    source_file: str | None = None
    source_file_hash: str | None = None
    source_ref: str | None = None
    imported_at: str | None = None
    modified_at: str | None = None
    calculation_method: str | None = None
    calculation_parameters: dict | None = None
    software_version: str | None = None
    verification_state: VerificationState
    notes: str | None = None
